### ISSUE TITLE
OAuth2::AccessToken.from_hash: `hash` contained more than one 'token' key warning

### ISSUE THREAD

**Issue by Adnilson** [oauth-xx/oauth2#628]
Hi,

We upgraded `omniauth-auth0` to version 3.0.0 in our Rails app and the following warning is on our logs all the time in all of our environments:

``OAuth2::AccessToken.from_hash: `hash` contained more than one 'token' key (["access_token", "id_token"]); using "access_token"``

From what I saw on this [change](https://github.com/oauth-xx/oauth2/pull/624/files#diff-b665e6fc2096be34f9c5e92cb0f38eb3a229da37ed4da92487f0a834eb6ae336R24) on PR 624.

We haven't changed anything in our app besides upgrading the gems associated as well (omniauth and omniauth-rails_csrf_protection) to their latest version.

Is there a way to solve this or circumvent?

EDIT: mind you that the `id_token` will be passed regardless: https://github.com/auth0/omniauth-auth0/blob/master/lib/omniauth/strategies/auth0.rb#L47

**Comment by pboling** [comment]
If the warning is saying a different token will be used than is actually used, that may be a bug, but it sounds like if it was it would break your use case.  Another PR just fixed an issue where this new code was selecting the wrong token: #626 and that fix is planned for release 2.0.7.

@Adnilson Are you able to test against master to make sure your auth code is still functional?  This code base has 100% test coverage on lines AND branches, so if something isn't working, it needs to be documented, fixed, and tests added.

There is currently no way to silence this warning.  If you can't control that too many tokens are being returned to you perhaps we need to add that as a feature, like `silence_extra_tokens_warning`?

**Comment by Adnilson** [comment]
Hi, @pboling thanks for the answer.

In that PR @rickselby had a problem where their login functionality broke, ours didn't.

We are not using the gem directly. It is through `omniauth-auth0`, and dependency wise it's working like this: `omniauth-auth0 -> omniauth-oauth2 -> oauth2`

I added `oauth2` to the Gemfile referencing master `(ref: 5e20fdd6016e617b5b3080652dc75c43342b2b5a)` and the warning is still there as nothing was modified on that area of the code.

I think the change has to be upwards on the dependency path.

**Comment by pboling** [comment]
I am not sure I understand the comment, but the warning is coming from this gem, so if we add a config, you can configure it, even though it is a nested dependency.  We may need to add a global config to support that use case, since you may not be able to config the client setup by the Omniauth gems.

**Comment by pboling** [comment]
My suggestion to test master is unrelated to the warning.  It changes which token key is used in some scenarios, but apparently didn't affect yours because, I am guessing, the strategy doesn't utilize the standardized token.

I will investigate further soon.  On parental leave right now.

**Comment by morgoth** [comment]
Getting the same warning, but we use the oauth2 gem directly like this, so looks like we're using the public methods of this gem. Not sure what's wrong with that:
```ruby
OAuth2::Client.new(
      ENV["GOOGLE_SIGN_IN_CLIENT_ID"],
      ENV["GOOGLE_SIGN_IN_CLIENT_SECRET"],
      authorize_url: "https://accounts.google.com/o/oauth2/auth",
      token_url: "https://oauth2.googleapis.com/token",
      redirect_uri: "some-redirect-url-here
    ).auth_code.get_token(params[:code])
```

**Comment by pboling** [comment]
@morgoth Do you have any idea why the response being parsed has multiple tokens in it?

**Comment by morgoth** [comment]
I don't know, but I confirm that the response from `https://oauth2.googleapis.com/token` has both access_token and id_token in the json

**Comment by pboling** [comment]
Containing both is triggering the warning... How should this gem decide which one to use?  Previously it went through a list of keys, and the first match got used.  But the ordering wasn't documented.

How would you suggest we improve this?

**Comment by morgoth** [comment]
I'm not very into internals of how this exactly works, so I cannot give any suggestions - sorry.
I just noticed a warning in CI and found this thread.

We have a code: `client.auth_code.get_token(params[:code])["id_token"]`, so we take what we need explicitly

**Comment by pboling** [comment]
Yeah, that's why it works "stably" for you.  The intended api is `client.auth_code.get_token(params[:code]).token`, but for that to work we have pick the right token... :/

**Comment by pboling** [comment]
There is a new option `token_name` you can use to specify the exact parameter name of the token you want to extract with `OAuth2::AccessToken`.  This avoids the warning altogether regardless of what other params are present.
